sdk/lib/opentelemetry/sdk/trace/trace_state.rb

```diff
@@ -0,0 +1,186 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Trace
+      # TraceState represents the OpenTelemetry::Trace::Tracestate values that
+      # are part of the OpenTelemetry ecosystem, contained in a single entry
+      # using the `ot` key. The format and handling requirements are specified
+      # at https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-handling.md
+      class TraceState
+        class << self
+          private :new # rubocop:disable Style/AccessModifierDeclarations
+
+          # Returns a newly created TraceState parsed from the Tracestate provided.
+          #
+          # @param [OpenTelemetry::Trace::Tracestate] tracestate
+          # @return [TraceState] A new TraceState instance or DEFAULT
+          def from_tracestate(tracestate) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+            return DEFAULT if tracestate.empty?
+
+            ot = tracestate.value('ot')
+            return DEFAULT if ot.nil? || ot.length > MAX_LIST_LENGTH
+
+            hash = ot.split(';').each_with_object({}) do |member, memo|
+              member.strip!
+              kv = member.split(':')
+              k, v = *kv
+              # TODO: "the used keys MUST be unique." - do we validate the header during parsing or just during formatting?
+              next unless kv.length == 2 && VALID_KEYS.include?(k) && VALID_VALUE.match?(v)
```

Comment:
```suggestion
              k, v, remainder = member.strip.split(':')
              # TODO: "the used keys MUST be unique." - do we validate the header during parsing or just during formatting?
              next unless remainder.nil? && VALID_KEYS.include?(k) && VALID_VALUE.match?(v)
```

```
irb(main):001:0> a = "foo:bar"
=> "foo:bar"
irb(main):002:0> k, v, remainder = a.strip.split(":")
=> ["foo", "bar"]
irb(main):003:0> remainder.nil?
=> true
irb(main):004:0> a = "foo:bar:baz"
=> "foo:bar:baz"
irb(main):005:0> k, v, remainder = a.strip.split(":")
=> ["foo", "bar", "baz"]
irb(main):006:0> remainder.nil?
=> false
irb(main):007:0> RUBY_VERSION
=> "2.7.2"
irb(main):008:0>
```

Reply:
I guess I didn't bother to demonstrate but `k` and `v` get destructured correctly in those examples, as well.

Reply:
For the `TODO` comment about uniqueness validation ... I'm not sure if we ever really need to do that?

Consider:

```
irb(main):006:1* hash = "foo:bar;foo:baz".split(";").each_with_object({}) do |member, memo|
irb(main):007:1*   k, v, remainder = member.strip.split(":")
irb(main):008:1*   next unless remainder.nil? # not doing the key or value validation for the example
irb(main):009:1*   memo[k] = v
irb(main):010:0> end
=> {"foo"=>"baz"}
irb(main):011:0>
```

Similar thoughts apply for the `from_hash` method as well; it would be very hard indeed to construct a hash that actually resulted in duplicate keys once serialized.

Reply:
The suggestion has one more allocation (for the `strip` vs `strip!`) per iteration.

Reply:
> The suggestion has one more allocation (for the strip vs strip!) per iteration.

And we do need to be careful with allocations, so ... never mind! 😄 

> so in this case it is very much possible to have duplicates in the input (but not in the output).

Ah yeah, good point. Maybe we just `warn` if we see a duplicate key and keep going.

> are we required to validate uniqueness when parsing the ot tracestate entry, and reject invalid entries, or do we just adopt "last one wins"?

😬 My vote is "warn during parse, allow last one to win" ... and maybe we should bring this up in the spec directly then?